multimap/src/main/java/org/infinispan/multimap/impl/function/PutFunction.java

```diff
@@ -0,0 +1,72 @@
+package org.infinispan.multimap.impl.function;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.infinispan.commons.api.functional.EntryView;
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.multimap.impl.ExternalizerIds;
+
+/**
+ * The adding function to remove a ket or a key/value on the multimap
+ *
+ * @author Katia Aresti - karesti@redhat.com
+ * @since 9.1
+ */
+public final class PutFunction<K, V> extends BaseFunction<K, V, Void> {
+
+   public static final AdvancedExternalizer<PutFunction> EXTERNALIZER = new Externalizer();
+   private final K key;
+   private final V value;
+
+   public PutFunction(K key, V value) {
+      this.key = key;
+      this.value = value;
+   }
+
+   @Override
+   public Void apply(EntryView.ReadWriteEntryView<K, Collection<V>> entryView) {
+      Collection<V> newValues = new HashSet<>();
+      newValues.add(value);
+
+      try {
+         // Raises NoSuchElementException if the value does not exist
+         Collection<V> values = entryView.get();
```

Comment:
Hrmm if the entry exists can't we just add the new value to it and set it back in the view? Otherwise the update will get progressively slower.